bindings/serialize.cc

```diff
@@ -0,0 +1,290 @@
+    : id(id),
+      nameX(nameX),
+      systemNameX(systemNameX),
+      filenameX(filenameX),
+      startLine(startLine) {}
+
+uint64_t ProfileFunction::getID() { return id; }
+int64_t ProfileFunction::getNameX() { return nameX; }
+int64_t ProfileFunction::getSystemNameX() { return systemNameX; }
+int64_t ProfileFunction::getFilenameX() { return filenameX; }
+int64_t ProfileFunction::getStartLine() { return startLine; }
+
+void ProfileFunction::encode(std::vector<char>* buffer) const {
+  encodeUint64Opt(1, id, buffer);
+  encodeInt64Opt(2, nameX, buffer);
+  encodeInt64Opt(3, systemNameX, buffer);
+  encodeInt64Opt(4, filenameX, buffer);
+  encodeInt64Opt(5, startLine, buffer);
+}
+
+ProfileLocation::ProfileLocation(uint64_t id, uint64_t mappingID,
+                                 uint64_t address, std::vector<Line> line,
+                                 bool isFolded)
+    : id(id),
+      mappingID(mappingID),
+      address(address),
+      line(line),
+      isFolded(isFolded) {}
+
+uint64_t ProfileLocation::getID() { return id; }
+
+uint64_t ProfileLocation::getMappingID() { return mappingID; }
+
+uint64_t ProfileLocation::getAddress() { return address; }
+
+std::vector<Line> ProfileLocation::getLine() { return line; }
+
+bool ProfileLocation::getIsFolded() { return isFolded; }
+
+void ProfileLocation::encode(std::vector<char>* buffer) const {
+  encodeUint64Opt(1, id, buffer);
+  encodeInt64Opt(2, mappingID, buffer);
+  encodeInt64Opt(3, address, buffer);
+  encodeRepeatedMessage<Line>(4, line, buffer);
+  encodeBoolOpt(5, isFolded, buffer);
+}
+
+Sample::Sample(std::vector<uint64_t> locationID, std::vector<int64_t> value,
+               std::vector<Label> label)
+    : locationID(locationID), value(value), label(label) {}
+
+std::vector<uint64_t> Sample::getLocationID() { return locationID; }
+
+std::vector<int64_t> Sample::getValue() { return value; }
+
+std::vector<Label> Sample::getLabel() { return label; }
+
+void Sample::encode(std::vector<char>* buffer) const {
+  encodeUint64s(1, locationID, buffer);
+  encodeInt64s(2, value, buffer);
+  encodeRepeatedMessage<Label>(3, label, buffer);
+}
+
+Profile::Profile(std::string periodType, std::string periodUnit, int64_t period,
+                 int64_t timeNanos, int64_t durationNanos,
+                 std::string dropFrames, std::string keepFramesX)
+    : period(period),
+      timeNanos(timeNanos),
+      durationNanos(durationNanos),
+      defaultSampleTypeX(0) {
+  // first index of strings must be ""
+  stringID("");
+  this->periodType = ValueType(stringID(periodType), stringID(periodUnit));
+  this->dropFramesX = stringID(dropFrames);
+  this->keepFramesX = stringID(keepFramesX);
+}
+
+void Profile::addSampleType(std::string type, std::string unit) {
+  int64_t typeX = stringID(type);
+  int64_t unitX = stringID(unit);
+  sampleType.push_back(ValueType(typeX, unitX));
+}
+
+void Profile::addSample(std::unique_ptr<Node>& node,
```

Comment:
Why does this accept a `unique_ptr<Node>&` rather than a `Node&`?

Reply:
This applies to other places as well.

Reply:
Or "const Node*"? I don't think Node is mutated here.

Reply:
Done.
Using const Node&